File: baekjoon_js/dfs&bfs/1167.js
("use strict");

const fs = require("fs");
const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `5
1 3 2 -1
2 4 4 -1
3 1 2 4 3 -1
4 2 4 3 3 5 6 -1
5 4 6 -1
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const N = Number(input());

let graph = Array(N + 1)
  .fill(null)
  .map(() => Array());

for (let i = 0; i < N; i++) {
  let nodeList = input().split(" ").map(Number);
  let startNode = nodeList[0];
  let j = 1;
  while (nodeList[j] != -1) {
    graph[startNode].push([nodeList[j], nodeList[j + 1]]);
    j += 2;
  }
}

// console.log(graph);

function bfs(start) {
  q = [];
  q.push([start, 0]);

  while (q.length > 0) {
    [node, value] = q.shift();
    isVisited[node] = true;
    // console.log(node, value);
    for ([nextNode, nextValue] of graph[node]) {
      //   console.log(nextNode, nextValue);
      if (isVisited[nextNode] === false) {
        q.push([nextNode, value + nextValue]);
        if (maxValue < value + nextValue) {
          //   console.log(maxNode, maxValue);
          maxValue = value + nextValue;
          maxNode = nextNode;
        }
      }
    }
  }
}

let [maxValue, maxNode] = [0, 0];
let isVisited = Array(N + 1).fill(false);
bfs(1);

maxValue = 0;
isVisited = Array(N + 1).fill(false);
bfs(maxNode);

console.log(maxValue);

File: backjoon_review/dfs&bfs/1167.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input())

graph = [[] * (N + 1) for _ in range(N + 1)]

for i in range(1, N + 1):
    inputList = list(map(int,input().split()))

    for i in range(1, len(inputList)-1, 2):
        graph[inputList[0]].append((inputList[i], inputList[i+1]))

def bfs(startNode):
    global maxValue, maxNode
    q = deque()
    q.append((startNode, 0))
    isVisited[startNode] = True
    while q:
        node, cost = q.popleft()

        for i in graph[node]:
            if isVisited[i[0]] == False:
                isVisited[i[0]] = True
                q.append((i[0], i[1] + cost))
                if maxValue < i[1] + cost:
                    maxValue = i[1] + cost
                    maxNode = i[0]

maxValue = 0
maxNode = 0
isVisited = [False] * (N + 1)
bfs(1)

maxValue = 0
isVisited = [False] * (N + 1)
bfs(maxNode)
print(maxValue)